x/reward/keeper/reward.go

```diff
@@ -0,0 +1,184 @@
+	blockRatioA := sdk.NewDec(int64(lastBlockHeight)).Sub(sdk.NewDec(int64(rewardPool.CurrentRewardHeight)))
+	blockRatioB := sdk.NewDec(int64(rewardPool.LastRewardHeight)).Sub(sdk.NewDec(int64(rewardPool.CurrentRewardHeight)))
+	blockRatio := blockRatioA.Quo(blockRatioB)
+	if blockRatio.GT(sdk.NewDec(1)) {
+		blockRatio = sdk.NewDec(1)
+	}
+
+	totalRelativeSignaturesDistributed := sdk.NewDec(0)
+	// distribute rewards to all block signers
+	for _, signatureCount := range signatureCounts.Counts {
+
+		// get the validator address from the cons address
+		// if the validator is not registered, reward distribution is skipped
+		// all funds are sent back to the coordinator
+		validatorByConsAddr, found := k.profileKeeper.GetValidatorByConsAddress(ctx, signatureCount.ConsAddress)
+		if found {
+			validator, found := k.profileKeeper.GetValidator(ctx, validatorByConsAddr.ValidatorAddress)
+			if !found {
+				return spnerrors.Criticalf(
+					"validator by consensus address not associated with a validator %s",
+					validatorByConsAddr.ValidatorAddress,
+				)
+			}
+			// calculate the total relative signature distributed for we calculate the refund
+			totalRelativeSignaturesDistributed = totalRelativeSignaturesDistributed.Add(signatureCount.RelativeSignatures)
+
+			// compute reward relative to the signature and block count
+			// and update reward pool
+			signatureRatio := signatureCount.RelativeSignatures.Quo(sdk.NewDec(int64(signatureCounts.BlockCount)))
+			reward, err := CalculateReward(blockRatio, signatureRatio, rewardPool.Coins)
+			if err != nil {
+				return spnerrors.Criticalf("invalid reward: %s", err.Error())
+			}
+			coins, isNegative := rewardPool.Coins.SafeSub(reward)
+			if isNegative {
+				return spnerrors.Criticalf("negative reward pool: %s", rewardPool.Coins.String())
+			}
+			rewardPool.Coins = coins
+
+			// send rewards to the address
+			account, err := sdk.AccAddressFromBech32(validator.Address)
+			if err != nil {
+				return spnerrors.Criticalf("can't parse address %s", err.Error())
+			}
+			if err := k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, account, reward); err != nil {
+				return spnerrors.Criticalf("send rewards error: %s", err.Error())
+			}
+		}
+	}
+	// if the reward pool is closed or last reward height is reached
+	// the remaining coins are refunded and reward pool is deleted
+	if closeRewardPool || lastBlockHeight >= rewardPool.LastRewardHeight {
+		if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+			ctx,
+			types.ModuleName,
+			provider,
+			rewardPool.Coins); err != nil {
+			return spnerrors.Criticalf("send rewards error: %s", err.Error())
+		}
+		k.RemoveRewardPool(ctx, launchID)
+		return nil
+	}
+
+	return k.RefundRewards(
+		ctx,
+		rewardPool,
+		provider,
+		signatureCounts,
+		totalRelativeSignaturesDistributed,
+		blockRatio,
+		launchID,
+		lastBlockHeight,
+		closeRewardPool,
+	)
+}
+
+func (k Keeper) RefundRewards(
```

Comment:
I'm ok reverting the change if you consider the method is more complex this way
If we let it like this we should document the method

Reply:
Yeah, I agree to revert, is more complex, and the `RefundRewards` is harder to test, too many args